mountpoint-s3-fs/src/metablock.rs

```diff
@@ -52,23 +56,47 @@ pub trait Metablock: Send + Sync {
     /// The kernel may forget a number of references (`n`) in one forget message to our FUSE implementation.
     async fn forget(&self, ino: InodeNo, n: u64);
 
-    /// Start writing to an inode.
-    async fn start_writing(&self, ino: InodeNo, mode: &WriteMode, is_truncate: bool) -> Result<(), InodeError>;
+    /// Open a new file handle for the given inode in read or write mode depending on flags and inode state.
+    async fn open_handle(
+        &self,
+        ino: InodeNo,
+        fh: u64,
+        write_mode: &WriteMode,
+        flags: OpenFlags,
+    ) -> Result<NewHandle, InodeError>;
 
     /// Increase the size of a file open for writing.
     /// Parameter `len` refers to the additional
     /// Returns the new size after the increase.
     async fn inc_file_size(&self, ino: InodeNo, len: usize) -> Result<usize, InodeError>;
 
-    /// Called when the filesystem has finished writing to the inode refernced by `ino`.
+    /// Called when the filesystem has finished writing to the inode referenced by `ino`.
     /// Allows the implementor to make necessary adjustments / update its internal structure.
-    async fn finish_writing(&self, ino: InodeNo, etag: Option<ETag>) -> Result<(), InodeError>;
-
-    /// Prepare an inode (referenced by `ino`) to start reading.
-    async fn start_reading(&self, ino: InodeNo) -> Result<(), InodeError>;
+    async fn finish_writing(&self, ino: InodeNo, etag: Option<ETag>, fh: u64) -> Result<Lookup, InodeError>;
```

Comment:
consider merging `finish_*` methods into one with an argument `mode: ReadWriteMode` specifying the type (the same for `flush_*`)

this would be consistent with `try_activate_handle` signature